From 689427102793caf6669bb041687821ab0fc35d10 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Tue, 27 Oct 2020 11:40:52 +0700
Subject: [PATCH 05/18] rzg: drivers: add delay driver

Refer from "rcar_gen3: drivers: delay: Rewrite from assembler to C"
to support delay driver written in C for rzg gen2 platform.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/renesas/rzg/delay/micro_delay.c | 30 ++++++++++++++++++++++++++++++
 drivers/renesas/rzg/delay/micro_delay.h | 15 +++++++++++++++
 2 files changed, 45 insertions(+)
 create mode 100644 drivers/renesas/rzg/delay/micro_delay.c
 create mode 100644 drivers/renesas/rzg/delay/micro_delay.h

diff --git a/drivers/renesas/rzg/delay/micro_delay.c b/drivers/renesas/rzg/delay/micro_delay.c
new file mode 100644
index 0000000..2209dc8
--- /dev/null
+++ b/drivers/renesas/rzg/delay/micro_delay.c
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <arch.h>
+#include <arch_helpers.h>
+#include "micro_delay.h"
+
+#define RZG_CONV_MICROSEC		1000000U
+
+void
+#if IMAGE_BL31
+	__attribute__ ((section (".system_ram")))
+#endif
+	rzg_micro_delay(uint64_t micro_sec)
+{
+	uint64_t freq;
+	uint64_t base_count;
+	uint64_t get_count;
+	uint64_t wait_time = 0U;
+
+	freq = read_cntfrq_el0();
+	base_count = read_cntpct_el0();
+	while (micro_sec > wait_time) {
+		get_count = read_cntpct_el0();
+		wait_time = ((get_count - base_count) * RZG_CONV_MICROSEC) / freq;
+	}
+}
diff --git a/drivers/renesas/rzg/delay/micro_delay.h b/drivers/renesas/rzg/delay/micro_delay.h
new file mode 100644
index 0000000..6130dfb
--- /dev/null
+++ b/drivers/renesas/rzg/delay/micro_delay.h
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MICRO_DELAY_H
+#define MICRO_DELAY_H
+
+#ifndef __ASSEMBLER__
+#include <stdint.h>
+void rzg_micro_delay(uint64_t micro_sec);
+#endif
+
+#endif /* MICRO_DELAY_H */
-- 
2.7.4

